lib/libzfs/libzfs_dataset.c

```diff
@@ -4028,6 +4028,25 @@ zfs_destroy_snaps_nvl(libzfs_handle_t *hdl, nvlist_t *snaps, boolean_t defer)
 			    dgettext(TEXT_DOMAIN, "snapshot is cloned"));
 			ret = zfs_error(hdl, EZFS_EXISTS, errbuf);
 			break;
+		case EBUSY: {
+			nvlist_t *existing_holds;
+			int err = lzc_get_holds(nvpair_name(pair),
+			    &existing_holds);
+
+			/* check the presence of holders */
+			if (err == 0 && !nvlist_empty(existing_holds)) {
+				const nvpair_t *holder = nvlist_next_nvpair(
+				    existing_holds, NULL);
+				zfs_error_aux(hdl,
+				    dgettext(TEXT_DOMAIN, "held by %s"),
+				    nvpair_name(holder));
+				ret = zfs_error(hdl, EBUSY, errbuf);
+				nvlist_free(existing_holds);
```

Comment:
In case of empty nvlist it is not freed.  I haven't looked to the roots of nvlist code, but it feels wrong.  It does not look like empty nvlist does not need free.